generator/src/genclient/translator/protobuf/protobuf.go

```diff
@@ -0,0 +1,277 @@
+package protobuf
+
+import (
+	"log/slog"
+	"strings"
+
+	"github.com/googleapis/google-cloud-rust/generator/src/genclient"
+	"github.com/googleapis/google-cloud-rust/generator/src/genclient/language"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+	"google.golang.org/protobuf/types/pluginpb"
+)
+
+type Translator struct {
+	request  *pluginpb.CodeGeneratorRequest
+	language string
+
+	// Only used for local testing
+	outDir      string
+	templateDir string
+
+	// State by FQN
+	state *genclient.APIState
+}
+
+type Options struct {
+	Request  *pluginpb.CodeGeneratorRequest
+	Language string
+	// Only used for local testing
+	OutDir      string
+	TemplateDir string
+}
+
+func NewTranslator(opts *Options) *Translator {
+	return &Translator{
+		outDir:      opts.OutDir,
+		request:     opts.Request,
+		language:    opts.Language,
+		templateDir: opts.TemplateDir,
+		state: &genclient.APIState{
+			ServiceByID: make(map[string]*genclient.Service),
+			MessageByID: make(map[string]*genclient.Message),
+			EnumByID:    make(map[string]*genclient.Enum),
+		},
+	}
+}
+
+// Translates proto representation into a [genclient.GenerateRequest].
+func (t *Translator) Translate() (*genclient.GenerateRequest, error) {
+	api := &genclient.API{
+		//TODO(codyoss): Read this from somewhere, likely service yaml or proto opt
+		Name: "secretmanager",
+	}
+	files := t.request.GetSourceFileDescriptors()
+	// TODO(codyoss): maybe need to process well known types: https://github.com/googleapis/gapic-generator-go/blob/main/internal/gengapic/well_known_types.go
+	for _, f := range files {
+		var fileServices []*genclient.Service
+		fFQN := "." + f.GetPackage()
+
+		// Messages
+		for _, m := range f.MessageType {
+			mFQN := fFQN + "." + m.GetName()
+			msg := t.processMessage(m, mFQN, nil)
+			api.Messages = append(api.Messages, msg)
+		}
+
+		// Services
+		for _, s := range f.Service {
+			service := &genclient.Service{}
+			sFQN := fFQN + "." + s.GetName()
+			t.state.ServiceByID[sFQN] = service
+			service.Name = s.GetName()
+			service.ID = sFQN
+			service.DefaultHost = parseDefaultHost(s.GetOptions())
+			for _, m := range s.Method {
+				method := &genclient.Method{}
+				method.HTTPInfo = parseHTTPInfo(m.GetOptions())
+				method.Name = m.GetName()
+				method.InputTypeID = m.GetInputType()
+				method.OutputTypeID = m.GetOutputType()
+				service.Methods = append(service.Methods, method)
+			}
+			fileServices = append(fileServices, service)
+		}
+
+		// Add docs
+		for _, loc := range f.GetSourceCodeInfo().GetLocation() {
+			p := loc.GetPath()
+			if loc.GetLeadingComments() == "" || len(p) == 0 {
+				continue
+			}
+
+			// These magic numbers come from reading the proto docs. They come
```

Comment:
A reference to the proto documentation would be useful.

Reply:
yeah, added a link